framework/src/Volo.Abp.IdentityModel/Volo/Abp/IdentityModel/IdentityModelAuthenticationService.cs

```diff
@@ -76,21 +81,36 @@ public virtual async Task<string> GetAccessTokenAsync(IdentityClientConfiguratio
                 throw new AbpException($"Could not retrieve the OpenId Connect discovery document! ErrorType: {discoveryResponse.ErrorType}. Error: {discoveryResponse.Error}");
             }
 
-            var tokenResponse = await GetTokenResponse(discoveryResponse, configuration);
-
-            if (tokenResponse.IsError)
+            var cacheKey = CalculateCacheKey(discoveryResponse, configuration);
+            var tokenCacheItem = await Cache.GetAsync(cacheKey);
+            if (tokenCacheItem == null)
             {
-                if (tokenResponse.ErrorDescription != null)
+                var tokenResponse = await GetTokenResponse(discoveryResponse, configuration);
+
+                if (tokenResponse.IsError)
                 {
-                    throw new AbpException($"Could not get token from the OpenId Connect server! ErrorType: {tokenResponse.ErrorType}. Error: {tokenResponse.Error}. ErrorDescription: {tokenResponse.ErrorDescription}. HttpStatusCode: {tokenResponse.HttpStatusCode}");
+                    if (tokenResponse.ErrorDescription != null)
+                    {
+                        throw new AbpException($"Could not get token from the OpenId Connect server! ErrorType: {tokenResponse.ErrorType}. " +
+                                               $"Error: {tokenResponse.Error}. ErrorDescription: {tokenResponse.ErrorDescription}. HttpStatusCode: {tokenResponse.HttpStatusCode}");
+                    }
+
+                    var rawError = tokenResponse.Raw;
+                    var withoutInnerException = rawError.Split(new string[] { "<eof/>" }, StringSplitOptions.RemoveEmptyEntries);
+                    throw new AbpException(withoutInnerException[0]);
                 }
 
-                var rawError = tokenResponse.Raw;
-                var withoutInnerException = rawError.Split(new string[] { "<eof/>" }, StringSplitOptions.RemoveEmptyEntries);
-                throw new AbpException(withoutInnerException[0]);
+                await Cache.SetAsync(cacheKey, new IdentityModelTokenCacheItem(tokenResponse.AccessToken),
+                    new DistributedCacheEntryOptions()
+                    {
+                        //Subtract 10 seconds of network request time.
+                        AbsoluteExpirationRelativeToNow = TimeSpan.FromSeconds(tokenResponse.ExpiresIn - 10)
```

Comment:
Should we remove the token from cache if we get an authentication/authorization related error from the server? If, somehow, server side invalidates the token, we continously get errors.

In addition, I think it is better to make the cache duration smaller, like 30 minutes for example (configurable), to force to re-authenticate the client. It would be better to refresh the token since client claims might be changed in this period.

Reply:
I added `CacheAbsoluteExpiration ` prorpery to `IdentityClientConfiguration`. You can continue to review.